### PR TITLE
Adds script to create the clusters for the nightly e2e jobs.

### PR DESCRIPTION
**Pull request by prashastia**
Adds script to create the clusters for the nightly e2e jobs.
1. gcloud command to create the temp, staging buckets and creating the cluster.
2. Adds the step in `nightly.yaml` for the cluster creation.
3. Call the above-mentioned script from `nightly.sh`.

/gcbrun

### CODE REVIEW

**Comment by prashastia** [comment]
Creates new clusters for all tests.
1. Small Bounded tests - includes small table read, nested schema read and query test.
2. Unbounded read- Reading the partitioned table with dynamic insertions 
3. Large Table Read - Reading a large table of O(100GB’s)


The 3 clusters are created concurrently

- Small Bounded Tests can be created in the regions - [asia-east1, asia-east2, asia-southeast1]
- Large Table Test can be created in the region [asia-northeast1, asia-northeast2, asia-northeast3]
- Unbounded read test can be created in the region [asia-south1, asia-south2, asia-southeast2]

Process: 
1. For each region the following steps are attempted:
- The staging and temp buckets are created in the region
- Then cluster creation is attempted
    - We wait for the staging and temp buckets to be successfully created as they take a variable amount of time to be created and be available for reading and writing.
    - Then, cluster creation is attempted.
    - In case cluster creation is successful then the region is set to a text file and the cluster name (which is CLUSTER_NAME+”_” + TIMESTAMP)
    - When cluster creation fails the same process is retried for another region in the arrays mentioned.
2. The cluster is created in one of the 3 regions.
3. Then, for each of the read tests we read the cluster name and the final region decided. 
4. The table read is tested.
5. Then after the success of last tests for each of the clusters (query-read for small table test, large and unbounded read have only one test) All the running jobs on the clusters are canceled, the staging and temp buckets are deleted and the cluster is also deleted.

File: cloudbuild/nightly/cloudbuild.yaml [hunk 1]
```diff
@@ -14,23 +14,78 @@ steps:
...
+    entrypoint: 'bash'
+    args: ['/workspace/cloudbuild/nightly/nightly.sh', 'create_clusters_bounded_small_table']
+    env:
+      - 'PROJECT_ID=${_PROJECT_ID}'
+      - 'DATAPROC_IMAGE_VERSION=${_DATAPROC_IMAGE_VERSION}'
+      - 'NUM_WORKERS_SMALL_TEST=${_NUM_WORKERS_SMALL_TEST}'
+      - 'CLUSTER_NAME_SMALL_TEST=${_CLUSTER_NAME_SMALL_TEST}'
+      - 'REGION_ARRAY_STRING_SMALL_TEST=${_REGION_ARRAY_STRING_SMALL_TEST}'
+#      - 'TEMP_BUCKET_SMALL_TEST=${_TEMP_BUCKET_SMALL_TEST}'
+#      - 'STAGING_BUCKET_SMALL_TEST=${_STAGING_BUCKET_SMALL_TEST}'
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/cloudbuild.yaml:30]: Why have these comments been added here?

Comment (jayehwhyehentee) [reply cloudbuild/nightly/cloudbuild.yaml:30]: ideal to remove them.

File: cloudbuild/nightly/cloudbuild.yaml [hunk 2]
```diff
@@ -14,23 +14,78 @@ steps:
...
+    entrypoint: 'bash'
+    args: ['/workspace/cloudbuild/nightly/nightly.sh', 'create_clusters_bounded_large_table']
+    env:
+      - 'PROJECT_ID=${_PROJECT_ID}'
+      - 'DATAPROC_IMAGE_VERSION=${_DATAPROC_IMAGE_VERSION}'
+      - 'NUM_WORKERS_LARGE_TABLE_TEST=${_NUM_WORKERS_LARGE_TABLE_TEST}'
+      - 'CLUSTER_NAME_LARGE_TABLE_TEST=${_CLUSTER_NAME_LARGE_TABLE_TEST}'
+      - 'REGION_ARRAY_STRING_LARGE_TABLE_TEST=${_REGION_ARRAY_STRING_LARGE_TABLE_TEST}'
+#      - 'TEMP_BUCKET_LARGE_TABLE_TEST=${_TEMP_BUCKET_LARGE_TABLE_TEST}'
+#      - 'STAGING_BUCKET_LARGE_TABLE_TEST=${_STAGING_BUCKET_LARGE_TABLE_TEST}'
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/cloudbuild.yaml:48]: Please remove.

File: cloudbuild/nightly/cloudbuild.yaml [hunk 3]
```diff
@@ -14,23 +14,78 @@ steps:
...
     entrypoint: 'bash'
+    args: ['/workspace/cloudbuild/nightly/nightly.sh', 'create_clusters_unbounded_table']
+    env:
+      - 'PROJECT_ID=${_PROJECT_ID}'
+      - 'DATAPROC_IMAGE_VERSION=${_DATAPROC_IMAGE_VERSION}'
+      - 'NUM_WORKERS_UNBOUNDED_TABLE_TEST=${_NUM_WORKERS_UNBOUNDED_TABLE_TEST}'
+      - 'CLUSTER_NAME_UNBOUNDED_TABLE_TEST=${_CLUSTER_NAME_UNBOUNDED_TABLE_TEST}'
+      - 'REGION_ARRAY_STRING_UNBOUNDED_TABLE_TEST=${_REGION_ARRAY_STRING_UNBOUNDED_TABLE_TEST}'
+#      - 'TEMP_BUCKET_UNBOUNDED_TABLE_TEST=${_TEMP_BUCKET_UNBOUNDED_TABLE_TEST}'
+#      - 'STAGING_BUCKET_UNBOUNDED_TABLE_TEST=${_STAGING_BUCKET_UNBOUNDED_TABLE_TEST}'
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/cloudbuild.yaml:66]: Remove

Comment (prashastia) [reply cloudbuild/nightly/cloudbuild.yaml]: Done.

File: cloudbuild/nightly/cloudbuild.yaml (hunk 2 above)
Comment (prashastia) [reply cloudbuild/nightly/cloudbuild.yaml]: Done.

File: cloudbuild/nightly/cloudbuild.yaml (hunk 1 above)
Comment (prashastia) [reply cloudbuild/nightly/cloudbuild.yaml]: Done.

File: cloudbuild/nightly/scripts/python-scripts/create_cluster.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        'project_id': project_id,
+        'cluster_name': cluster_name,
+        'config': {
+            'master_config': {'num_instances': 1, 'machine_type_uri': 'n1-standard-2'},
+            'worker_config': {'num_instances': num_workers, 'machine_type_uri': 'n2-standard-4'},
+            'software_config': {
+                'image_version': dataproc_image_version,
+                'optional_components': ['FLINK']},
+            'initialization_actions': [{'executable_file': initialisation_action_script_uri}],
+            'lifecycle_config': {'auto_delete_ttl': '3600s'},
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/scripts/python-scripts/create_cluster.py:53]: Is this enough? Whats the total duration of our nightly job?

File: cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    logging.info(f'Delete all the buckets connected to cluster {cluster_name}.')
+
+
+def delete_bucket(bucket_name):
+    logging.info(f'Attempting to delete the bucket {bucket_name}')
+    os.system(f'gcloud storage rm --recursive gs://{bucket_name}')
+    logging.info(f'Bucket {bucket_name} deleted')
+
+
+def delete_cluster(project_id, cluster_name, region):
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py:100]: Why delete the cluster explicitly if it is being created with a ttl?

File: cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+
+def clear_all_dataproc_dependencies(project_id, cluster_name, region):
+    delete_active_jobs(project_id, cluster_name, region)
+    delete_cluster_buckets(project_id, cluster_name, region)
+    delete_cluster(project_id, cluster_name, region)
+
+
+def delete_active_jobs(project_id, cluster_name, region):
+    """Function to list and cancel all running jobs on dataproc cluster.
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py:30]: Lets keep consistent names. The job is being canceled, not deleted.
Also, and more importantly, why do you need to explicitly cancel the job? Didnt you add an async timer on the Flink application itself to end at some point?

Comment (jayehwhyehentee) [reply cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py:30]: I see https://github.com/GoogleCloudDataproc/flink-bigquery-connector/blob/main/flink-connector-bigquery-integration-test/src/main/java/com/google/cloud/flink/bigquery/integration/BigQueryIntegrationTest.java#L225

File: cloudbuild/nightly/scripts/python-scripts/create_cluster.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    for region in region_array:
+        logging.info(f'Attempting cluster creation with region {region}')
+        is_cluster_created = create_cluster(project_id, region, cluster_name, num_workers,
+                                            dataproc_image_version,
+                                            initialisation_action_script_uri)
+        if is_cluster_created:
+            file = open(region_saving_file, 'w')
+            file.write(region)
+            file.close()
+            break
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/scripts/python-scripts/create_cluster.py:140]: Lets mention in comments what exactly is being done here. Elaborate on why we're creating clusters in separate regions (resource quota limitations), and what is our approach here to tackle that.

File: cloudbuild/nightly/scripts/python-scripts/create_cluster.py (hunk 4 above)
Comment (prashastia) [reply cloudbuild/nightly/scripts/python-scripts/create_cluster.py:53]: The nightly job (including cluster creation) runs for about 31mins.

File: cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py (hunk 6 above)
Comment (prashastia) [reply cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py]: yes. Makes Sense. Fixed.

File: cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+
+def delete_bucket(bucket_name):
+    logging.info(f'Attempting to delete the bucket {bucket_name}')
+    os.system(f'gcloud storage rm --recursive gs://{bucket_name}')
+    logging.info(f'Bucket {bucket_name} deleted')
+
+
+def delete_cluster(project_id, cluster_name, region):
+    """Function to list and cancel all running jobs on dataproc cluster.
```
[the remark this answers is not in the document]
Comment (prashastia) [reply cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py]: Fixed.

File: cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py (hunk 5 above)
Comment (jayehwhyehentee) [reply cloudbuild/nightly/scripts/python-scripts/delete_buckets_and_clusters.py:59]: Makes sense. I agree its better to delete as soon as tests are over.

File: cloudbuild/nightly/nightly.sh [hunk 9]
```diff
@@ -28,38 +28,98 @@ case $STEP in
...
+    timestamp=$(date +"%Y%m%d%H%M%S")
+    # 1. Create the first cluster for bounded read.
+    # - Modify the cluster name for all tests.
+    CLUSTER_NAME_SMALL_TEST="$CLUSTER_NAME_SMALL_TEST"-"$timestamp"
+    # - call script that creates cluster with retries.
+    source cloudbuild/nightly/scripts/create_dataproc_cluster.sh "$CLUSTER_NAME_SMALL_TEST" "$REGION_ARRAY_STRING_SMALL_TEST" "$NUM_WORKERS_SMALL_TEST" "$REGION_SMALL_TEST_FILE" "$WORKER_MACHINE_TYPE_SMALL_BOUNDED"
+    # - save the cluster for future uses
+    echo "$CLUSTER_NAME_SMALL_TEST" > "$CLUSTER_SMALL_TEST_FILE"
+    exit
+    ;;
```
Comment (jayehwhyehentee) [inline cloudbuild/nightly/nightly.sh:43]: I see a template which can be defined as one function and called with relevant arguments. There is code repetition right now.

```
  create_cluster)
    timestamp=$(date +"%Y%m%d%H%M%S")
    CLUSTER_NAME="$CLUSTER_NAME"-"$timestamp"
    source cloudbuild/nightly/scripts/create_dataproc_cluster.sh "$CLUSTER_NAME" "$REGION_ARRAY_STRING" "$NUM_WORKERS" "$REGION_NAME_TEST_NAME_FILE" "$WORKER_MACHINE_TYPE"
    echo "$CLUSTER_NAME_TEST_NAME" > "$CLUSTER_NAME_TEST_NAME_FILE"
    exit
    ;;
```

Comment (jayehwhyehentee) [reply cloudbuild/nightly/nightly.sh:43]: Program arguments ($FOO) can be converted to method arguments when invoking this function.

**Review by jayehwhyehentee — APPROVED** [review]

**Pull request merged**